flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/conversion/HiveInspectors.java

```diff
@@ -181,15 +191,72 @@ public static HiveObjectConversion getConversion(ObjectInspector inspector, Data
 					inspector instanceof BinaryObjectInspector) {
 				return IdentityConversion.INSTANCE;
 			} else if (inspector instanceof HiveCharObjectInspector) {
-				return o -> new HiveChar((String) o, ((CharType) dataType.getLogicalType()).getLength());
+				return o -> new HiveChar((String) o, ((CharType) dataType).getLength());
 			} else if (inspector instanceof HiveVarcharObjectInspector) {
-				return o -> new HiveVarchar((String) o, ((VarCharType) dataType.getLogicalType()).getLength());
+				return o -> new HiveVarchar((String) o, ((VarCharType) dataType).getLength());
 			}
 
 			// TODO: handle decimal type
 		}
 
-		// TODO: handle complex types like struct, list, and map
+		if (inspector instanceof ListObjectInspector) {
+			HiveObjectConversion eleConvert = getConversion(
+				((ListObjectInspector) inspector).getListElementObjectInspector(),
+				((ArrayType) dataType).getElementType());
+			return o -> {
+				Object[] array = (Object[]) o;
+				List<Object> result = new ArrayList<>();
```

Comment:
We can probably instantiate the list with the known size.

Reply:
unfortunately we don't know the size because array data can be variable size.